bindings/python/tests/test_nfa.py

```diff
@@ -274,6 +465,108 @@ def test_intersection(
     assert mata.Nfa.is_included(inter, fa_one_divisible_by_four, alph)[0]
     assert map == {(0,0): 0, (1,1): 1, (1,3): 3, (2,2): 2, (2, 4): 4}
 
+
+def test_intersection_epsilon_preserving():
+    epsilon = ord('e')
+    a = mata.Nfa(6)
+    a.make_initial_state(0)
+    a.make_final_states([1, 4, 5])
+    a.add_trans_raw(0, epsilon, 1)
+    a.add_trans_raw(1, ord('a'), 1)
+    a.add_trans_raw(1, ord('b'), 1)
+    a.add_trans_raw(1, ord('c'), 2)
+    a.add_trans_raw(2, ord('b'), 4)
+    a.add_trans_raw(2, epsilon, 3)
+    a.add_trans_raw(3, ord('a'), 5)
+
+    b = mata.Nfa(10)
+    b.make_initial_state(0)
+    b.make_final_states([2, 4, 8, 7])
+    b.add_trans_raw(0, ord('b'), 1)
+    b.add_trans_raw(0, ord('a'), 2)
+    b.add_trans_raw(2, ord('a'), 4)
+    b.add_trans_raw(2, epsilon, 3)
+    b.add_trans_raw(3, ord('b'), 4)
+    b.add_trans_raw(0, ord('c'), 5)
+    b.add_trans_raw(5, ord('a'), 8)
+    b.add_trans_raw(5, epsilon, 6)
+    b.add_trans_raw(6, ord('a'), 9)
+    b.add_trans_raw(6, ord('b'), 7)
+
+    result, prod_map = mata.Nfa.intersection_epsilon_preserving(a, b, epsilon)
+
+    # Check states.
+    for i, m in prod_map.items():
```

Comment:
I'd say this is useful if some of the test fails, so I would remove it right now, since it will only taint the output of pytest.

Reply:
That was a debugging print I have forgotten to remove. Fixed.

However, the bug I was debugging here might interest you. The previous Python binding function `intersection()` returned the following product map (with the resulting product automaton): 
```python
return result, {tuple(k): v for k, v in product_map}
```
For tuples of original states `(x, y)` where `x < y`, the Python object for product map is constructed correctly. However, for tuples where `x > y`, the original state names (numbers) are sorted and from tuple `(1, 0)` in libMata, we create a pair `(0, 1)` in Python. I updated this bug in this function, but there is one other occurrence of this pattern: `subset_map_to_dictionary`. Could you look into that to make sure this does not produce unexpected results? I suppose this was a copy-paste error when the original Python binding function for intersection was implemented, but I want to be sure this is not a problem there, too.

Reply:
Make a mental (or some other) note, that we should discuss this on Thursday. There is actually a different reasoning for the sorting and I see, why this is problem for the product (naturally, the order matters here), however, for subsets I don't think the order matters (those are macrostates, which are sets; the sorting is performed so the resulting tuple is deterministic, i.e. the same everytime the code is run; the issue is that sets cannot be hashed, hence we convert and use the tuples instead).

Reply:
From what I have seen in the function for subsets, I believe it should not matter, indeed. Nevertheless, I will make a note and mention this on Thursday.